### ISSUE TITLE
Javascript Bug when using jQuery.DataTables

### ISSUE THREAD

**Issue by patrickdobsonGL** [divio/django-cms#6016]
### Summary
The Plugin Picker does not populate correctly when [jQuery.DataTables](https://datatables.net/) is used on the page.

### Expected behaviour
jQuery.DataTables is loaded and initialized on the page. When switching to edit mode and clicking the "+" button to add a new plugin to the Structure view, the Plugin Picker should populate the modal and display a list of the available plugins:

![image](https://user-images.githubusercontent.com/28363678/28083992-13aac330-662c-11e7-8dcd-e995546b1751.png)

#### Actual behaviour
The plugin picker doesn't display any plugins.

![forad-304-broken](https://user-images.githubusercontent.com/28363678/28085054-59db2a90-662f-11e7-936c-bb16b10060d7.png)

No Javascript errors seem to accompany this bug.

#### Environment

* Python version: 2.7
* Django version: 1.8
* django CMS version: 3.4.0

JQuery.Data Tables initialization:

HTML:

    <table id="my-table">
        <thead>
            <tr>
                <th class="no-sort"></th>
                <th>Name</th>
                <th>Category</th>
                <th>Type</th>
                <th>Description</th>
                <th class="no-sort"></th>
            </tr>
        </thead>
        <tbody>
            {% for item in items %}
                <tr id="item-row-{{ item.pk }}">
                    <td></td>
                    <td>{{ item.name }}</td>
                    <td>{{ item.category }}</td>
                    <td>{{ item.type }}</td>
                    <td>{{ item.description }}</td>
                    <td></td>
                </tr>
            {% endfor %}
        </tbody>
    </table>

Javascript
`$('#my-table').DataTable({});`

**Comment by vxsx** [comment]
Hi, thanks for the report. CMS uses it's own jQuery for a while now and we don't expose it, so the data tables plugin shouldn't really affect things. However, it might be rewriting the markup in a way that this script tag goes missing for that particular placeholder https://github.com/divio/django-cms/blob/d49b5449af00c3d5fcff2daad195f2f5edf6920c/cms/templates/cms/toolbar/placeholder.html#L6-L8

would you mind sharing fuller source of the page (with relevant bits being the placeholder that's broken and the cms script tags and whatnot)?

**Comment by patrickdobsonGL** [comment]
Hey, thanks for the reply!

This is what is rendered for that snippet on the Structure for the given placeholder:

    <div class="cms-placeholder cms-placeholder-2">
    </div>
    <script id="cms-plugin-child-classes-2" type="text/cms-template">
        <div class="cms-submenu-item cms-submenu-item-title">
            <span>Applications</span>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="ExcelPlugin">Excel Report</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="TableauPlugin">Tableau Report</a>
        </div>
        <div class="cms-submenu-item cms-submenu-item-title">
            <span>Generic</span>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="FilePlugin">File</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="FolderPlugin">Folder</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="GoogleMapPlugin">Google Map</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="PicturePlugin">Image</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="LinkPlugin">Link</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="SnippetPlugin">Snippet</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="StylePlugin">Style</a>
        </div>
        <div class="cms-submenu-item">
             <a data-rel="add" href="TextPlugin">Text</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="VideoPlayerPlugin">Video player</a>
        </div>
        <div class="cms-submenu-item cms-submenu-item-title">
             <span>Google Map</span>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="GoogleMapMarkerPlugin">Marker</a>
        </div>
        <div class="cms-submenu-item">
             <a data-rel="add" href="GoogleMapRoutePlugin">Route</a>
        </div>
        <div class="cms-submenu-item cms-submenu-item-title">
             <span>Multi Columns</span>
        </div>
        <div class="cms-submenu-item">
             <a data-rel="add" href="ColumnPlugin">Column</a>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="MultiColumnPlugin">Multi Columns</a>
        </div>
        <div class="cms-submenu-item cms-submenu-item-title">
             <span>Video player</span>
        </div>
        <div class="cms-submenu-item">
            <a data-rel="add" href="VideoSourcePlugin">Source</a>
        </div>
        <div class="cms-submenu-item">
             <a data-rel="add" href="VideoTrackPlugin">Track</a>
        </div>
    </script>

**Comment by vxsx** [comment]
@patrickdobsonGL right, and when data-tables is initialized - is that script still there in the DOM? You can check by opening the console and querying: `CMS.$('#cms-plugin-child-classes-2').length`

**Comment by patrickdobsonGL** [comment]
That query returns 0.

After closer DOM inspection, it looks like DataTables is removing the entire <script> tag from the DOM when it initializes. When DataTables is removed, the <script> tag stays on the page and the plugin picker renders properly.

I suppose I could grab that script tag and move it somewhere else before DataTables initializes. I'll try something like that and report back with my results.

As for Django CMS, is the placeholder the best place to put data for the Structure view in edit mode? It seems like any DOM manipulation of placeholder elements opens up the possibility of breaking the CMS. If there's a way to separate data for the user-facing view and the structure view, that might help avoid problems like this in the future.

Also, there wasn't much by way of documentation for digging into the Django CMS Javascript modules. I probably could have debugged this on my own if there was a simple way to switch between the human readable Javascript and the uglified version - pretty printing in the developer's console always helps, but without proper variable names, it's still quite a chore.

**Comment by patrickdobsonGL** [comment]
Here's the workaround I'm using:

Before I initialize DataTables, I initialize a variable for the script tag and check to see if we're in draft mode:

    var scriptTag;
    if (CMS.config.mode === 'draft') {

Then I detach the table's CMS plugins script tag:

    scriptTag = $('#table-id tbody script');
    scriptTag.detach();

Then I initialize datatables:

    $('#table-id').DataTables({});

After datatables is initialized, I re-attach the script tag:
    if(CMS.config.mode === 'draft') {
        $('#table-id tbody').prepend($scriptTag);
    }

**Comment by vxsx** [comment]
Normally that script tag and the `cms-placeholder` div is on the same level as plugins so I've never actually seen it being affected. However, it's a valid point, we should move them away from the content, since in our 3.5 efforts their position is no longer important /cc @czpython 

> If there's a way to separate data for the user-facing view and the structure view, that might help avoid problems like this in the future.

There is going to be.

As for the JS/CSS - there is only this at the moment http://docs.django-cms.org/en/release-3.4.x/contributing/code.html#frontend - it mentions `gulp --debug` that would include source maps. Of course you would have to install cms from a local location for it to work.
